pallets/gated-marketplace/src/functions.rs

```diff
@@ -150,9 +151,334 @@ impl<T: Config> Pallet<T> {
+
+        //ensure the marketplace exists
+        ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+
+        //ensure the collection exists
+        //For this case user doesn't have to be the owner of the collection
+        //but the owner of the item cannot create a buy offer for their own collection
+        if let Some(a) = pallet_uniques::Pallet::<T>::owner(collection_id, item_id) {
+            ensure!(a != authority, Error::<T>::CannotCreateOffer);
+        } else {
+            Err(Error::<T>::CollectionNotFound)?;
+        }
+
+        //ensure user has enough balance to create the offer
+        let total_user_balance = T::LocalCurrency::total_balance(&authority);
+        ensure!(total_user_balance >= price, Error::<T>::NotEnoughBalance);
+
+        //ensure the price is valid
+        Self::is_the_price_valid(price)?;
+
+        //Add timestamp to the offer
+        let(timestamp, timestamp2) = Self::get_timestamp_in_milliseconds().ok_or(Error::<T>::TimestampError)?;
+
+        //create an offer_id
+        let offer_id = (marketplace_id, authority.clone(), collection_id, timestamp, timestamp2).using_encoded(blake2_256);
+
+        //create offer structure
+        let offer_data = OfferData::<T> {
+            marketplace_id,
+            collection_id,
+            item_id,
+            creator: authority.clone(),
+            price,
+            creation_date: timestamp,
+            expiration_date: timestamp2,
+            status: OfferStatus::Open,
+            offer_type: OfferType::BuyOrder,
+            buyer: None,
+        };
+
+        //insert in OffersByItem
+        //An item can receive multiple buy offers
+        <OffersByItem<T>>::try_mutate(collection_id, item_id, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        //insert in OffersByAccount
+        <OffersByAccount<T>>::try_mutate(authority, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        //insert in OffersInfo
+        // ensure the offer_id doesn't exist
+        ensure!(!<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferAlreadyExists);
+        <OffersInfo<T>>::insert(offer_id, offer_data);
+
+        //Insert in OffersByMarketplace
+        <OffersByMarketplace<T>>::try_mutate(marketplace_id, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+    
+        Self::deposit_event(Event::OfferStored(collection_id, item_id));
+        Ok(())
+    }
+
+    pub fn do_take_sell_offer(buyer: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32]) -> DispatchResult {
+        //This extrisicn is called by the user who wants to buy the item
+        //ensure offer exists
+        ensure!(<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferNotFound);
```

Comment:
Is this line required, if you are checking that the offer exists when you get the offer data on line 294?

Reply:
Done.